mozilla-release/browser/components/sessionstore/SessionStore.jsm

```diff
@@ -487,6 +487,66 @@ var SessionStore = {
 // Freeze the SessionStore object. We don't want anyone to modify it.
 Object.freeze(SessionStore);
 
+const cliqz_removeDuplicatedEntries = function(tabs, overwriteTabs) {
+  // CLIQZ-SPECIAL: DB-2447
+  // overwriteTabs is always true if "Show Home Page" option is
+  // not selected on Preferences page.
+  // Otherwise overwriteTabs equals false.
+  //
+  // To handle deduplication problem we need to loop through
+  // every single tab and remove any of its' entries which
+  // should not be restored.
+  // What is left afterwards (if any) will be restored later on.
+  // If overwriteTabs is true then we need to make sure there is
+  // at least one entry left after squashing duplicates.
+  // To detect that we need to find the last entry that not be
+  // restored and then get an index of the tab it belongs to.
+  const homePages = HomePage.get().split("|");
+  const shouldNotRestoreEntry = function(entry) {
+    if (homePages.includes(entry.url)) {
+      return true;
+    }
+
+    return homePages.includes(HomePage.getOriginalDefault()) ?
+      entry.isCliqzPage === 1 :
+      false;
+  };
+
+  let lastTabIndex = tabs.length - 1;
+  if (overwriteTabs) {
+    lastTabIndex = (function() {
+      let i = lastTabIndex;
+      while (i >= 0) {
+        let entries = tabs[i].entries || [];
+
+        for (let j = 0, l = entries.length; j < l; j++) {
```

Comment:
From my understanding entries is the stack of navigation URLs for a given tab, we would only be interested is last entry, for restore decision,